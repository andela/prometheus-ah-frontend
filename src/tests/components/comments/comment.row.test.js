import React from 'react';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import ConnectedCommentRow, { CommentRow } from '../../../components/comment/CommentRow';
import mockData from '../../__mocks__/commentMockData';

const props = {
  createCommentReplyAction: jest.fn(),
  deleteCommentAction: jest.fn(),
  updateCommentAction: jest.fn(),
  loadCommentReplyAction: jest.fn(),
  article: mockData.articles[0],
  comment: mockData.comments.comments[0],
  auth: true,
  error: {},
  username: 'faksam',
  commentReplies: [
    {
      id: 2,
      createdAt: '2016-02-18T03:22:56.637Z',
      updatedAt: '2016-02-18T03:48:35.824Z',
      body: 'This article is hot',
      User: {
        username: 'joeeasy',
        email: 'joeeasy@gmail.com',
        bio: 'tomorrow is a better day',
        id: 1,
        isVerified: true,
        role: 'user'
      }
    },
    {
      id: 3,
      createdAt: '2016-02-18T03:22:56.637Z',
      updatedAt: '2016-02-18T03:48:35.824Z',
      body: 'This article is hot',
      User: {
        username: 'joeeasy',
        email: 'joeeasy@gmail.com',
        bio: 'tomorrow is a better day',
        id: 1,
        isVerified: true,
        role: 'user'
      }
    },
    {
      id: 4,
      createdAt: '2016-02-18T03:22:56.637Z',
      updatedAt: '2016-02-18T03:48:35.824Z',
      body: 'This article is hot',
      User: {
        username: 'joeeasy',
        email: 'joeeasy@gmail.com',
        bio: 'tomorrow is a better day',
        id: 1,
        isVerified: true,
        role: 'user'
      }
    },
  ],
};
const initialState = {
  articleReducer: {
    article: mockData.articles[0],
  },
  auth: {
    error: {}
  },
  commentReducer: {
    reply: mockData.commentReplies[0],
    commentReplies: mockData.commentReplies
  },

};
const mockStore = configureMockStore([thunk]);
const store = mockStore(initialState);

describe('CommentRow Component', () => {
  it('should render without throwing an error', () => {
    const wrapper = shallow(<CommentRow {...props} />);
    expect(wrapper.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });
  it('should post a comment update', () => {
    const wrapper = shallow(<CommentRow {...props} />);
    const instance = wrapper.instance();
    const event = {
      preventDefault: jest.fn()
    };
    wrapper.setState({ editBody: 'Samuel is a boy' });
    instance.isValid();
    instance.onSubmit(event);
    expect(props.updateCommentAction).toHaveBeenCalled();
  });
  it('should post a comment update', () => {
    const wrapper = shallow(<CommentRow {...props} />);
    const instance = wrapper.instance();
    const event = {
      preventDefault: jest.fn()
    };
    wrapper.setState({ editBody: '' });
    instance.isValid();
    instance.onSubmit(event);
    expect(props.updateCommentAction).toHaveBeenCalled();
  });
  it('should post a comment reply', () => {
    const wrapper = shallow(<CommentRow {...props} />);
    const instance = wrapper.instance();
    const event = {
      preventDefault: jest.fn()
    };
    wrapper.setState({ editBody: 'Samuel is a boy' });
    instance.isValid();
    instance.onReplySubmit(event);
    expect(props.createCommentReplyAction).toHaveBeenCalled();
  });
  it('should delete a comment reply', () => {
    const wrapper = shallow(<CommentRow {...props} />);
    const instance = wrapper.instance();
    const event = {
      preventDefault: jest.fn()
    };
    instance.handleCommentDelete(event);
    expect(props.deleteCommentAction).toHaveBeenCalled();
  });
  it('should start editing a comment reply', () => {
    const wrapper = shallow(<CommentRow {...props} />);
    const instance = wrapper.instance();
    const event = {
      preventDefault: jest.fn()
    };
    instance.handleCommentEdit(event);
    expect(wrapper.state().editing).toEqual(true);
  });
  it('should cancel comment reply editing', () => {
    const wrapper = shallow(<CommentRow {...props} />);
    const instance = wrapper.instance();
    const event = {
      preventDefault: jest.fn()
    };
    instance.handleCommentEditCancel(event);
    expect(wrapper.state().editing).toEqual(false);
  });
  it('should display reply editing form', () => {
    const wrapper = shallow(<CommentRow {...props} />);
    const instance = wrapper.instance();
    const event = {
      preventDefault: jest.fn()
    };
    instance.handleCommentReply(event);
    expect(wrapper.state().newReply).toEqual(true);
  });
  it('should hide reply editing form', () => {
    const wrapper = shallow(<CommentRow {...props} />);
    const instance = wrapper.instance();
    const event = {
      preventDefault: jest.fn()
    };
    instance.handleReplyCancel(event);
    expect(wrapper.state().newReply).toEqual(false);
  });
  it('should set state to props', () => {
    const wrapper = shallow(<ConnectedCommentRow store={store} {...props} />);
    expect(wrapper.state()).toEqual({});
  });

  it('should set password when password changes', () => {
    const wrapper = shallow(<CommentRow {...props} />);
    const event = {
      preventDefault: jest.fn(),
      target: {
        id: 'editBody',
        value: 'This is a comment'
      }
    };
    wrapper.setState({ errors: { [event.target.id]: 'mock' } });
    wrapper.instance().onChange(event);
    expect(wrapper.instance().state[event.target.id])
      .toBe(event.target.value);
  });
});
