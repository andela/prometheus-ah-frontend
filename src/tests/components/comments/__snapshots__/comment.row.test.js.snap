// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentRow Component should render without throwing an error 1`] = `
<div>
  <div
    className="card card-inner"
  >
    <div>
      <div
        className="row card-body"
      >
        <div
          className="col-2"
        >
          <img
            alt="user-thumbnail"
            className="rounded-circle img img-fluid-comment"
            src={Object {}}
          />
        </div>
        <div
          className="col-10"
        >
          <div
            className="comment-body"
          >
            <p
              id="username"
            >
              <strong>
                joeeasy
              </strong>
              <span
                className="comment-span"
                id="timestamp"
              >
                  -  
                3 years ago
              </span>
            </p>
            <p>
              This article is hot
            </p>
            <p
              className="comment-span"
            >
              <button
                className="btn btn-link"
                onClick={[Function]}
                type="button"
              >
                Reply
              </button>
              <span
                className="comment-span"
                id="threadCount"
              />
              -  
              <i
                className="mdi mdi-thumb-up-outline mr-3 align-top"
              >
                <span
                  className="likes-num"
                >
                   30
                </span>
              </i>
            </p>
            <div>
              <Connect(CommentReplyList)
                comment={
                  Object {
                    "User": Object {
                      "bio": "tomorrow is a better day",
                      "email": "joeeasy@gmail.com",
                      "id": 1,
                      "isVerified": true,
                      "role": "user",
                      "username": "joeeasy",
                    },
                    "body": "This article is hot",
                    "createdAt": "2016-02-18T03:22:56.637Z",
                    "id": 1,
                    "updatedAt": "2016-02-18T03:48:35.824Z",
                  }
                }
                replies={Array []}
                username="faksam"
              />
            </div>
          </div>
        </div>
        <div
          className="pull-right"
        >
           
        </div>
      </div>
    </div>
  </div>
  <br />
  <br />
</div>
`;
